File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;

    static int N;
    static Node[] tree;

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(br.readLine());
        tree = new Node[N];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            char data = st.nextToken().charAt(0);
            char left = st.nextToken().charAt(0);
            char right = st.nextToken().charAt(0);
            tree[data - 'A'] = new Node(data, left, right);
        }
        preorder(tree[0]);
        bw.write("\n");
        inorder(tree[0]);
        bw.write("\n");
        postorder(tree[0]);
        bw.flush();
        bw.close();
    }

    static void preorder(Node node) throws IOException {
        bw.write(node.data);
        if (node.left != '.') preorder(tree[node.left - 'A']);
        if (node.right != '.') preorder(tree[node.right - 'A']);
    }

    static void inorder(Node node) throws IOException {
        if (node.left != '.') inorder(tree[node.left - 'A']);
        bw.write(node.data);
        if (node.right != '.') inorder(tree[node.right - 'A']);
    }

    static void postorder(Node node) throws IOException {
        if (node.left != '.') postorder(tree[node.left - 'A']);
        if (node.right != '.') postorder(tree[node.right - 'A']);
        bw.write(node.data);
    }

    static class Node {
        char data;
        char left;
        char right;

        public Node(char data, char left, char right) {
            this.data = data;
            this.left = left;
            this.right = right;
        }
    }
}

File: 백준/Silver/1991. 트리 순회/트리 순회.py
class Node:
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right

def insertNode(temp, root, left, right):
    if temp.value == root:
        temp.left = Node(left) if left != '.' else None
        temp.right = Node(right) if right != '.' else None
    else:
        if temp.left is not None:
            insertNode(temp.left, root, left, right)
        if temp.right is not None:
            insertNode(temp.right, root, left, right)

def preOrder(node):
    if node is None:
        return
    print(node.value, end='')
    preOrder(node.left)
    preOrder(node.right)

def inOrder(node):
    if node is None:
        return
    inOrder(node.left)
    print(node.value, end='')
    inOrder(node.right)

def postOrder(node):
    if node is None:
        return
    postOrder(node.left)
    postOrder(node.right)
    print(node.value, end='')

n = int(input())
head = Node('A')

for _ in range(n):
    root, left, right = input().split()
    insertNode(head, root, left, right)

preOrder(head)
print()
inOrder(head)
print()
postOrder(head)
print()